RevenueCatUI/Templates/V2/Components/Carousel/CarouselComponentView.swift

```diff
@@ -0,0 +1,693 @@
+                            msTimePerSlide: style.autoAdvance?.msTimePerPage,
+                            msTransitionTime: style.autoAdvance?.msTransitionTime
+                        ).clipped()
+                    }
+                    // Need to set height since geometry reader has no intrinsic height
+                    .frame(height: carouselHeight)
+                    .onPreferenceChange(HeightPreferenceKey.self) { newHeight in
+                        self.carouselHeight = newHeight
+                    }
+                    // Style the carousel
+                    .size(style.size)
+                    .padding(style.padding)
+                    .shape(border: style.border,
+                           shape: style.shape,
+                           background: style.backgroundStyle,
+                           uiConfigProvider: self.viewModel.uiConfigProvider)
+                    .shadow(shadow: style.shadow, shape: style.shape?.toInsettableShape())
+                    .padding(style.margin)
+                } else {
+                    EmptyView()
+                }
+            }
+        }
+    }
+
+}
+
+struct HeightPreferenceKey: PreferenceKey {
+    static var defaultValue: CGFloat = 0
+    static func reduce(value: inout CGFloat, nextValue: () -> CGFloat) {
+        value = max(value, nextValue())
+    }
+}
+
+/// A wrapper to give each page copy a stable, unique identity.
+private struct CarouselItem<Content: View>: Identifiable {
+    let id: Int
+    let view: Content
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+private struct CarouselView<Content: View>: View {
+    // MARK: - Configuration
+
+    private let width: CGFloat
+    private let initialIndex: Int
+    private let originalPages: [Content]
+    private let loop: Bool
+    private let spacing: CGFloat
+    private let cardWidth: CGFloat
+
+    private let pageControl: DisplayablePageControl?
+
+    /// Optional auto-play timings (in milliseconds).
+    private let msTimePerSlide: Int?
+    private let msTransitionTime: Int?
+
+    // Number of pages in the user's original set.
+    private var originalCount: Int { originalPages.count }
+
+    // MARK: - State
+
+    /// The “expanded” data array, each item with a unique ID.
+    @State private var data: [CarouselItem<Content>] = []
+
+    /// The current index (in `data`) of the "active" page.
+    @State private var index: Int = 0
+
+    /// Real-time drag offset from the user’s finger.
+    @GestureState private var translation: CGFloat = 0
+
+    /// A timer for auto-play, if enabled.
+    @State private var autoTimer: Timer?
+    @State private var isPaused: Bool = false
+    @State private var pauseEndDate: Date?
+
+    // MARK: - Init
+
+    init(
+        width: CGFloat,
+        pages: [Content],
+        initialIndex: Int,
+        loop: Bool = false,
+        spacing: CGFloat = 16,
+        cardWidth: CGFloat = 300,
```

Comment:
Hmm seems weird to have these defaults? 🤔 Seems these should be in the previews if needed?

Reply:
I was going to comment the same. If we're passing values anyways, we should avoid the defaults. I've been burned in the past by unnecessary default values

Reply:
Ooooops... this was here from my POC before I had an API figured out 😅